### ISSUE TITLE
Cannot handle regexps with Japanese

### ISSUE THREAD

**Issue by rubikitch** [BurntSushi/ripgrep#309]
$ cd /tmp
    $ wget -q -O japanese.org http://rubikitch.com/f/japanese.org
    $ cat japanese.org
    * <2017-01-09 Mon>日本語タイトル
    Hello!
    $ ag --file-search-regex .\*.org --nocolor --line-number --smart-case --nogroup --column -- \^\(\\\*\)\+.\*\日\本\語 japanese.org
    1:1:* <2017-01-09 Mon>日本語タイトル
    $ rg  --no-heading --vimgrep \^\(\\\*\)\+.\*\日\本\語 japanese.org
    $ rg  --no-heading --vimgrep \日\本\語 japanese.org
    1:19:* <2017-01-09 Mon>日本語タイトル
    $ rg --version
    ripgrep 0.3.2

**Comment by BurntSushi** [comment]
I can't reproduce the issue. Namely, I see this:

```
$ rg  --no-heading --vimgrep \^\(\\\*\)\+.\*\日\本\語 japanese.org
1:1:* <2017-01-09 Mon>日本語タイトル
```

With that said, I find your regex very puzzling and hard to read. Why are you mixing shell escaping with regex escaping? It's almost guaranteed to lead to sadness. For example, this is much clearer:

```
$ rg '^(\*)+.*日本語' japanese.org
1:* <2017-01-09 Mon>日本語タイトル
```

To me, this suggests your problem is with escapes. Namely, if `^` is escaped inside the regex, then it will be treated as a literal `^`, which will cause your specific search to fail. But if `^` is escaped by the shell, then it will be interpreted as the special begin-line assertion. For example, this search fails:

```
$ rg '\^(\*)+.*日本語' japanese.org
```

I see no reason why the Japanese characters have anything at all to do with this bug. ripgrep's Unicode support is strictly superior to ag's.

**Comment by rubikitch** [comment]
I cannot reproduce it. What's wrong?
The complex regex is generated by org-seek.el/ripgrep.el in Emacs (shell-quote-argument).

    $ cd /tmp
    $ rg --no-heading --vimgrep '日本語' japanese.org
    1:19:* <2017-01-09 Mon>日本語タイトル
    $ rg --no-heading --vimgrep \^\(\\\*\)\+.\*\日\本\語 japanese.org
    $ rg --no-heading --vimgrep '^(\*)+.*日本語' japanese.org

It works well in non-Japanese chars.

    $ rg --no-heading --vimgrep \^\(\\\*\)\+.\*Mon japanese.org
    1:1:* <2017-01-09 Mon>日本語タイトル
    $ rg --no-heading --vimgrep '^(\*)+.*Mon' japanese.org
    1:1:* <2017-01-09 Mon>日本語タイトル

**Comment by BurntSushi** [comment]
I'm stumped on this one. I've even specifically tried ripgrep `0.3.2` (as opposed to current master), and I still can't reproduce the bug you're seeing. It honestly doesn't make any sense to me. Is there anything else about your environment that might be impacting search? What shell are you running? (Although, I can't even think of something that would cause this bug.)

If you're still willing to debug this, can you try to widdle down the regex that fails? What is the *smallest* regex you can write with the `日本語` literal that still fails? For example, does `.*日本語` work? What about `^.*日本語`? Can you try others?

**Comment by rubikitch** [comment]
This problem is fixed in 0.4.0. Thank you very much!!

**Comment by BurntSushi** [comment]
@rubikitch Strange. I still don't understand what was going on or what could have caused the bug. :-/

If anyone can reproduce it and do the [minimization steps](https://github.com/BurntSushi/ripgrep/issues/309#issuecomment-271458479), that would be much appreciated. Until then, I'll close this.
